File: 4챕터/movingAverage/movingAverage.ts
// M * (N-M+1) -> N*M-M**2+M 알고리즘
const movingAverage1 = (numberArray: number[], M: number): number[] => {
  const tempArray: number[] = [];

  for (let i = M - 1; i < numberArray.length; ++i) {
    let partialSum = 0;
    for (let j = 0; j < M; ++j) {
      partialSum += numberArray[i - j];
    }
    tempArray.push(partialSum / M);
  }

  return tempArray;
};

// N 알고리즘 (linear time algorithm)
const movingAverage2 = (numberArray: number[], M: number) => {
  const tempArray: number[] = [];
  let partialSum = 0;
  numberArray.forEach((currentValue: number, index: number) => {
    partialSum += currentValue;

    if (index >= M - 1) {
      tempArray.push(partialSum / M);
      partialSum -= numberArray[index - M + 1];
    }
  });
  return tempArray;
};

const test = [1, 2, 3, 4, 5, 6, 7, 8, 9];
console.log(movingAverage1(test, 4));
console.log(movingAverage2(test, 4));

File: 4챕터/movingAverage/movingAverage.cpp
// M * (N-M+1) -> N*M-M**2+M 알고리즘
vector<double> movingAverage1(const vector<double>& A, int M){
    vector<double> ret;
    int N = A.size();
    for(int i=M-1; i<N; ++i){
        double partialSum = 0;
        for(int j=0; j<M; ++j){
            partialSum += A[i-j];
        }
        ret.push_back(partialSum / M);
    }
    return ret;
}

// N 알고리즘 -> linear time 알고리즘
vector<double> movingAverage2(const vector<double>& A, int M){
    vector<double> ret;
    int N = A.size;
    double partialSum = 0;
    for(int i=0; i<M-1; ++i){
        partialSum += A[i];
    }
    for(int i=M-1; i<N; ++i){
        partialSum += A[i];
        ret.push_back(partialSum / M);
        partialSum -= A[i-M+1];
    }
    return ret;
}